### ISSUE TITLE
HyperKube image host rate limit causing deployments to fail.

### ISSUE THREAD

**Issue by jklauser** [Azure/acs-engine#3537]
**Is this a request for help?**:

--- YES

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

--- ISSUE

**What version of acs-engine?**:

--- 0.20.0

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes 1.11

**What happened**:
Our deployments have been failing this morning because the default Hyperkube image server in the East US region, https://k8s-gcrio.azureedge.net/v2/hyperkube-amd64/blobs/sha256:7460c29cea530abb10d39c00e52718293cd0969fed8d5f581cff38fd2d6bd550, is rate limiting requests.  Hitting it from other locations, i.e. my laptop, works fine.

The initial error is shown as:

```
New-AzureRmResourceGroupDeployment : 11:25:37 - Resource Microsoft.Compute/virtualMachines/extensions 'k8s-master-XXXXXXXX-0/cse-master-0' failed with message '{
  "status": "Failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'Failed'.",
    "details": [
      {
        "code": "VMExtensionProvisioningError",
        "message": "VM has reported a failure when processing extension 'cse-master-0'. Error message: \"Enable failed: failed to execute command: command terminated with exit status=31\n[stdout]\n\n[stderr]\n\"."
      }
    ]
  }
}'
```

Looking at /var/log/azure/cluster-provision.log we see 75 lines of:
```
Pulling k8s-gcrio.azureedge.net/hyperkube-amd64:v1.11.0...
httpReaderSeeker: failed open: unexpected status code https://k8s-gcrio.azureedge.net/v2/hyperkube-amd64/blobs/sha256:7460c29cea530abb10d39c00e52718293cd0969fed8d5f581cff38fd2d6bd550: 429 Too Many Requests
```
before it finally fails.

Looking at https://github.com/Azure/acs-engine/blob/master/parts/k8s/kubernetescustomscript.sh the default behavior is to retry 75 times.
`retrycmd_if_failure 75 1 60 img pull $HYPERKUBE_URL || exit $ERR_K8S_DOWNLOAD_TIMEOUT`

Looking at https://github.com/Azure/acs-engine/blob/master/parts/k8s/kubernetesprovisionsource.sh and given the line above, the retrycmd_if_failure function will sleep for one second between attempts.

**What you expected to happen**:
The deployment of the cluster is successful.

We can edit the configuration and modify the URL for this resource to bypass this issue but given how critical some of these components are to a deployment, it may be worthwhile to explore some options to ensure successful deployments when this situation arises.  For example:

- Have a progressive backoff in the script when receiving a 429 error to reduce the load on a server that may consider itself under attack to prevent it the server from remaining in a rate limiting mode for long periods of time.  (Over four hours now.)
- Have a facility in the script for primary and secondary locations to handle a situation where the primary server is unavailable.

**How to reproduce it** (as minimally and precisely as possible):
Deploy a new Kubernetes 1.11 cluster in East US

**Anything else we need to know**:
Deployments using ACS directly, e.g. AZ ACS CREATE, are successful.

**Comment by jackfrancis** [comment]
Sorry for the fire drill today, folks, the fix is in master and just patched as v0.20.1:

https://github.com/Azure/acs-engine/releases/tag/v0.20.1

tl;dr Putting these image pulls behind a CDN had the unintended side-effect of taking AKS and acs-engine distributed traffic sources and combining them into a few IP addresses, which (in eastus) was inducing 429 responses from gcr.io (which our CDN was not designed to mask).

Hence, we are no longer employing that CDN layer and sourcing directly from gcr.io!

**Comment by caseywatson** [comment]
Thanks, @jackfrancis!

**Comment by jklauser** [comment]
Thank you!  Clusters are being created again.
